File: app/src/main/java/com/checkin/app/checkin/Auth/AuthRepository.java
package com.checkin.app.checkin.Auth;

import android.app.Application;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;

import com.checkin.app.checkin.Data.ApiClient;
import com.checkin.app.checkin.Data.ApiResponse;
import com.checkin.app.checkin.Data.BaseRepository;
import com.checkin.app.checkin.Data.Converters;
import com.checkin.app.checkin.Data.NetworkBoundResource;
import com.checkin.app.checkin.Data.Resource;
import com.checkin.app.checkin.Data.RetrofitLiveData;
import com.checkin.app.checkin.Data.WebApiService;
import com.fasterxml.jackson.databind.node.ObjectNode;

import javax.inject.Singleton;

import retrofit2.Call;

@Singleton
public class AuthRepository extends BaseRepository {
    private static AuthRepository INSTANCE;
    private final WebApiService mWebService;

    private AuthRepository(Context context) {
        mWebService = ApiClient.Companion.getApiService(context);
    }

    public static AuthRepository getInstance(Application application) {
        if (INSTANCE == null) {
            synchronized (AuthRepository.class) {
                if (INSTANCE == null) {
                    INSTANCE = new AuthRepository(application.getApplicationContext());
                }
            }
        }
        return INSTANCE;
    }

    public LiveData<Resource<AuthResultModel>> login(final ObjectNode credentials) {
        return new NetworkBoundResource<AuthResultModel, AuthResultModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<AuthResultModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.postLogin(credentials));
            }

            @Override
            protected void saveCallResult(AuthResultModel data) {
            }
        }.getAsLiveData();
    }

    public LiveData<Resource<AuthResultModel>> register(final ObjectNode credentials) {
        return new NetworkBoundResource<AuthResultModel, AuthResultModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<AuthResultModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.postRegister(credentials));
            }

            @Override
            protected void saveCallResult(AuthResultModel data) {
            }
        }.getAsLiveData();
    }

    public Call<ObjectNode> postDeviceToken(String token) {
        ObjectNode data = Converters.INSTANCE.getObjectMapper().createObjectNode();
        data.put("device_token", token);
        return mWebService.postFCMToken(data);
    }
}

File: app/src/main/java/com/checkin/app/checkin/auth/AuthRepository.kt
package com.checkin.app.checkin.auth

import android.app.Application
import android.content.Context
import androidx.lifecycle.LiveData
import com.checkin.app.checkin.data.BaseRepository
import com.checkin.app.checkin.data.Converters.objectMapper
import com.checkin.app.checkin.data.network.ApiClient.Companion.getApiService
import com.checkin.app.checkin.data.network.ApiResponse
import com.checkin.app.checkin.data.network.RetrofitLiveData
import com.checkin.app.checkin.data.network.WebApiService
import com.checkin.app.checkin.data.resource.NetworkBoundResource
import com.checkin.app.checkin.data.resource.Resource
import com.checkin.app.checkin.utility.SingletonHolder
import com.fasterxml.jackson.databind.node.ObjectNode
import retrofit2.Call

class AuthRepository private constructor(context: Context) : BaseRepository() {
    private val mWebService: WebApiService = getApiService(context)

    fun authenticate(credentials: AuthRequestModel): LiveData<Resource<AuthResultModel>> {
        return object : NetworkBoundResource<AuthResultModel, AuthResultModel>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<AuthResultModel>> = RetrofitLiveData(mWebService.postAuthenticate(credentials))

            override fun saveCallResult(data: AuthResultModel?) {}
        }.asLiveData
    }

    fun postDeviceToken(token: String?): Call<ObjectNode> {
        val data = objectMapper.createObjectNode()
        data.put("device_token", token)
        return mWebService.postFCMToken(data)
    }

    companion object : SingletonHolder<AuthRepository, Application>({ AuthRepository(it.applicationContext) })
}